pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CaseTransformFunction.java

```diff
@@ -0,0 +1,482 @@
+            case DOUBLE:
+            case STRING:
+              dataType = resultMetadata.getDataType();
+              break;
+            default:
+              throw new IllegalStateException(
+                  String.format("Incompatible expression types in THEN Clause [%s].", resultMetadata));
+          }
+          break;
+        case LONG:
+          switch (resultMetadata.getDataType()) {
+            case INT:
+            case LONG:
+              break;
+            case FLOAT:
+            case DOUBLE:
+              dataType = FieldSpec.DataType.DOUBLE;
+              break;
+            case STRING:
+              dataType = FieldSpec.DataType.STRING;
+              break;
+            default:
+              throw new IllegalStateException(
+                  String.format("Incompatible expression types in THEN Clause [%s].", resultMetadata));
+          }
+          break;
+        case FLOAT:
+          switch (resultMetadata.getDataType()) {
+            case INT:
+            case FLOAT:
+              break;
+            case LONG:
+            case DOUBLE:
+              dataType = FieldSpec.DataType.DOUBLE;
+              break;
+            case STRING:
+              dataType = resultMetadata.getDataType();
+              break;
+            default:
+              throw new IllegalStateException(
+                  String.format("Incompatible expression types in THEN Clause [%s].", resultMetadata));
+          }
+          break;
+        case DOUBLE:
+          switch (resultMetadata.getDataType()) {
+            case INT:
+            case FLOAT:
+            case LONG:
+            case DOUBLE:
+              break;
+            case STRING:
+              dataType = resultMetadata.getDataType();
+              break;
+            default:
+              throw new IllegalStateException(
+                  String.format("Incompatible expression types in THEN Clause [%s].", resultMetadata));
+          }
+          break;
+        case STRING:
+          switch (resultMetadata.getDataType()) {
+            case INT:
+            case FLOAT:
+            case LONG:
+            case DOUBLE:
+            case STRING:
+              break;
+            default:
+              throw new IllegalStateException(
+                  String.format("Incompatible expression types in THEN Clause [%s].", resultMetadata));
+          }
+          break;
+        default:
+          if (resultMetadata.getDataType() != dataType) {
+            throw new IllegalStateException(
+                String.format("Incompatible expression types in THEN Clause [%s].", resultMetadata));
+          }
+      }
+    }
+    return new TransformResultMetadata(dataType, true, false);
+  }
+
+  @Override
+  public TransformResultMetadata getResultMetadata() {
+    return _resultMetadata;
+  }
+
+  private int[] getSelectedArray(ProjectionBlock projectionBlock) {
+    int[] selected = new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
```

Comment:
So this is gathering the bitmap by invoking the BinaryOperatorTransformFunction?

Reply:
selected array contains the value of ELSE(0)/THEN(1-N) index per record after evaluating the WHEN clause.

Reply:
Can you please add some comments on this method?